sensor/common/detector/baseline/baseline.go

```diff
@@ -88,3 +102,170 @@ func (w *baselineEvaluator) IsOutsideLockedBaseline(pi *storage.ProcessIndicator
+	oe.lock.Lock()
+	defer oe.lock.Unlock()
+
+	containerMap := oe.deploymentBaselines[id]
+	if containerMap == nil {
+		return
+	}
+
+	// Decrement reference counts for all process sets used by this deployment
+	for _, contentHash := range containerMap {
+		oe.removeReference(contentHash)
+	}
+
+	delete(oe.deploymentBaselines, id)
+	log.Debugf("Successfully removed deployment %s from baseline", id)
+}
+
+// AddBaseline adds a process baseline to the optimized evaluator
+func (oe *optimizedBaselineEvaluator) AddBaseline(baseline *storage.ProcessBaseline) {
+	oe.lock.Lock()
+	defer oe.lock.Unlock()
+
+	deploymentID := baseline.GetKey().GetDeploymentId()
+	containerName := baseline.GetKey().GetContainerName()
+
+	// Check if baseline is locked (has UserLockedTimestamp)
+	if baseline.GetUserLockedTimestamp() == nil {
+		// Unlocked baseline - remove any existing baseline (processes are always considered within unlocked baselines)
+		if oe.deploymentBaselines[deploymentID] != nil {
+			if oldContentHash, exists := oe.deploymentBaselines[deploymentID][containerName]; exists {
+				oe.removeReference(oldContentHash)
+				delete(oe.deploymentBaselines[deploymentID], containerName)
+				if len(oe.deploymentBaselines[deploymentID]) == 0 {
+					delete(oe.deploymentBaselines, deploymentID)
+				}
+			}
+		}
+		log.Debugf("Successfully added unlocked process baseline %s", baseline.GetId())
+		return
+	}
+
+	// Locked baseline - process normally
+	baselineSet := set.NewStringSet()
+	for _, elem := range baseline.GetElements() {
+		if process := elem.GetElement().GetProcessName(); process != "" {
+			baselineSet.Add(process)
+		}
+	}
+
+	// Find existing process set with same content or create new one
+	contentHash := oe.findOrCreateProcessSet(baselineSet)
+
+	// Update deployment mapping
+	if oe.deploymentBaselines[deploymentID] == nil {
+		oe.deploymentBaselines[deploymentID] = make(map[string]string)
+	}
+
+	// If this deployment/container already has a process set, decrement its ref count
+	if oldContentHash, exists := oe.deploymentBaselines[deploymentID][containerName]; exists {
+		oe.removeReference(oldContentHash)
+	}
+
+	oe.deploymentBaselines[deploymentID][containerName] = contentHash
+	log.Debugf("Successfully added locked process baseline %s", baseline.GetId())
+}
+
+// findOrCreateProcessSet finds an existing process set with the same content or creates a new one
+func (oe *optimizedBaselineEvaluator) findOrCreateProcessSet(processes set.StringSet) string {
+	// Compute content hash for O(1) lookup
+	contentHash := computeProcessSetHash(processes)
+
+	// Check if we already have this process set
+	if entry, exists := oe.processSets[contentHash]; exists {
+		// COLLISION PROTECTION: Verify content actually matches
+		if !entry.processes.Equal(processes) {
+			log.Fatalf("CRITICAL: Hash collision detected for process set %v vs existing %v",
+				processes.AsSlice(), entry.processes.AsSlice())
+		}
```

Comment:
I think, we may want to use the soft-panic here. I mean that variant that panics in dev and only errors in prod.